src/mono/mono/metadata/native-library.c

```diff
@@ -519,18 +522,27 @@ netcore_probe_for_module (MonoImage *image, const char *file_name, int flags)
 	// If the difference becomes a problem, overhaul this algorithm to match theirs exactly
 
 	// Try without any path additions
-	module = netcore_probe_for_module_variations (NULL, file_name, lflags);
+	module = netcore_probe_for_module_variations (NULL, file_name, lflags, error);
+	if (!module && !is_ok (error) && mono_error_get_error_code (error) == MONO_ERROR_BAD_IMAGE)
+		return NULL;
 
 	// Check the NATIVE_DLL_SEARCH_DIRECTORIES
-	for (int i = 0; i < pinvoke_search_directories_count && module == NULL; ++i)
-		module = netcore_probe_for_module_variations (pinvoke_search_directories[i], file_name, lflags);
+	for (int i = 0; i < pinvoke_search_directories_count && module == NULL; ++i) {
+		mono_error_cleanup (error);
+		error_init_reuse (error);
+		module = netcore_probe_for_module_variations (pinvoke_search_directories[i], file_name, lflags, error);
+		if (!module && !is_ok (error) && mono_error_get_error_code (error) == MONO_ERROR_BAD_IMAGE)
```

Comment:
Yes, had that in mind as well, but since only Windows will trigger a MONO_ERROR_BAD_IMAGE at the moment, I didn't pursue it further. I can take a quick look to see if we can postpone the error and if the test is still passing doing it that way.

Reply:
I made a change where we will keep any last MONO_ERROR_BAD_IMAGE error (using mono_error_move), but continue probing. If we exhaust all probing attempts but still haven't been able to load a module, and have encountered a MONO_ERROR_BAD_IMAGE failure in any of the probing attempts, we will raise that error.

I believe this will be inline how it worked before, we will exhaust all probing attempts and complete as soon as we successfully loaded a module found in the probing paths. If we exhaust all probing attempts, but have hit a MONO_ERROR_BAD_IMAGE in any of the attempts, we will raise that exception (inline with the expectations of src\tests\Interop\NativeLibrary\AssemblyLoadContext\ResolveUnmanagedDllTests.csproj), if not we will raise the regular DLL not found exception.

Reply:
Thanks!

And thanks for reminding me about `mono_error_move`.